onnxruntime/core/framework/utils.h

```diff
@@ -87,69 +87,399 @@ void DumpNodeInputs(const OpKernelContext& context, const Node& node);
+  return ONNX_NAMESPACE::TensorProto_DataType_UINT32;
+};
+template <>
+constexpr ONNX_NAMESPACE::TensorProto_DataType ToTensorProtoElementType<uint64_t>() {
+  return ONNX_NAMESPACE::TensorProto_DataType_UINT64;
+};
+template <>
+constexpr ONNX_NAMESPACE::TensorProto_DataType ToTensorProtoElementType<BFloat16>() {
+  return ONNX_NAMESPACE::TensorProto_DataType_BFLOAT16;
+};
+
+template <typename T>
+constexpr ONNXTensorElementDataType GetONNXTensorElementDataType() {
+  return ONNX_TENSOR_ELEMENT_DATA_TYPE_UNDEFINED;
+}
+
+template <>
+constexpr ONNXTensorElementDataType GetONNXTensorElementDataType<bool>() {
+  return ONNX_TENSOR_ELEMENT_DATA_TYPE_BOOL;
+}
+
+template <>
+constexpr ONNXTensorElementDataType GetONNXTensorElementDataType<std::string>() {
+  return ONNX_TENSOR_ELEMENT_DATA_TYPE_STRING;
+}
+
+template <>
+constexpr ONNXTensorElementDataType GetONNXTensorElementDataType<float>() {
+  return ONNX_TENSOR_ELEMENT_DATA_TYPE_FLOAT;
+}
+
+template <>
+constexpr ONNXTensorElementDataType GetONNXTensorElementDataType<double>() {
+  return ONNX_TENSOR_ELEMENT_DATA_TYPE_DOUBLE;
+}
+
+template <>
+constexpr ONNXTensorElementDataType GetONNXTensorElementDataType<MLFloat16>() {
+  return ONNX_TENSOR_ELEMENT_DATA_TYPE_FLOAT16;
+}
+
+template <>
+constexpr ONNXTensorElementDataType GetONNXTensorElementDataType<BFloat16>() {
+  return ONNX_TENSOR_ELEMENT_DATA_TYPE_BFLOAT16;
+}
+
+template <>
+constexpr ONNXTensorElementDataType GetONNXTensorElementDataType<int8_t>() {
+  return ONNX_TENSOR_ELEMENT_DATA_TYPE_INT8;
+}
+
+template <>
+constexpr ONNXTensorElementDataType GetONNXTensorElementDataType<uint8_t>() {
+  return ONNX_TENSOR_ELEMENT_DATA_TYPE_UINT8;
+}
+
+template <>
+constexpr ONNXTensorElementDataType GetONNXTensorElementDataType<int16_t>() {
+  return ONNX_TENSOR_ELEMENT_DATA_TYPE_INT16;
+}
+
+template <>
+constexpr ONNXTensorElementDataType GetONNXTensorElementDataType<uint16_t>() {
+  return ONNX_TENSOR_ELEMENT_DATA_TYPE_UINT16;
+}
+
+template <>
+constexpr ONNXTensorElementDataType GetONNXTensorElementDataType<int32_t>() {
+  return ONNX_TENSOR_ELEMENT_DATA_TYPE_INT32;
+}
+
+template <>
+constexpr ONNXTensorElementDataType GetONNXTensorElementDataType<uint32_t>() {
+  return ONNX_TENSOR_ELEMENT_DATA_TYPE_UINT32;
+}
+
+template <>
+constexpr ONNXTensorElementDataType GetONNXTensorElementDataType<int64_t>() {
+  return ONNX_TENSOR_ELEMENT_DATA_TYPE_INT64;
+}
+
+template <>
+constexpr ONNXTensorElementDataType GetONNXTensorElementDataType<uint64_t>() {
+  return ONNX_TENSOR_ELEMENT_DATA_TYPE_UINT64;
+}
+
+// The following primitives are strongly recommended for switching on tensor input datatypes for
+// kernel implementations.
+//
+//  1) If you need to handle all of the primitive tensor contained datatypes, the best choice would be macros
+//     DispatchOnTensorType or DispatchOnTensorTypeWithReturn. Use inline wrappers so your function can be invoked as function<T>().
```

Comment:
could we name the macros DISPATCH_ON_TENSOR_TYPE and DISPATCH_ON_TENSOR_TYPE_WITH_RETURN?

Reply:
I have done several rounds of refactoring and I think if I did this the costs would outweigh the benefits.

---
In reply to: [343339194](https://github.com/microsoft/onnxruntime/pull/2307#discussion_r343339194) [](ancestors = 343339194)